content/03_CreatingObservableSequences.md

```diff
@@ -0,0 +1,1274 @@
+
+### From IEnumerable&lt;T&gt;
+
+Rx defines another extension method called `ToObservable`, this time for `IEnumerable<T>`. In earlier chapters I described how `IObservable<T>` was designed to represent the same basic abstraction as `IEnumerable<T>`, with the only difference being the mechanism we use to obtain the elements in the sequence: with `IEnumerable<T>`, we write code that _pulls_ values out of the collection (e.g., a `foreach` loop), whereas `IObservable<T>` _pushes_ values to us by invoking `OnNext` on our `IObserver<T>`.
+
+We could write code that bridges from _pull_ to _push_:
+
+```csharp
+// Example code only - do not use!
+public static IObservable<T> ToObservableOversimplified<T>(this IEnumerable<T> source)
+{
+    return Observable.Create<T>(o =>
+    {
+        foreach (var item in source)
+        {
+            o.OnNext(item);
+        }
+
+        o.OnComplete();
+
+        // Incorrectly ignoring unsubscription.
+        return Disposable.Empty;
+    });
+}
+```
+
+This crude implementation conveys the basic idea, but it is naive. It does not attempt to handle unsubscription, and it's not easy to fix that when using `Observable.Create` for this particular scenario. And as we will see later in the book, Rx sources that might try to deliver large numbers of events in quick succession should integrated with Rx's concurrency model. The implementation that Rx supplies does of course cater for all of these tricky details. That makes it rather more complex, but that's Rx's problem; you can think of it as being logically equivalent to the code shown above, but without the shortcomings.
+
+When transitioning from `IEnumerable<T>` to `IObservable<T>`, you should carefully consider what you are really trying to achieve. Consider that the blocking synchronous (pull) nature of `IEnumerable<T>` does always not mix well with the asynchronous (push) nature of `IObservable<T>`. As soon as something subscribes to an `IObservable<T>` created in this way, it is effectively asking to iterate over the `IEnumerable<T>`, immediately producing all of the values. The call to `Subscribe` might not return until it has reached the end of the `IEnumerable<T>`, making it similar to the very simple example shown [at the start of this chapter](#a-very-basic-iobservablet-implementation). (I say "might" because as we'll see when we get to schedulers, the exact behaviour depends on the context.) `ToObservable` can't work magic—something somewhere has to execute what amounts to a `foreach` loop.
+
+So although this can be a convenient way to bring sequences of data into an Rx world, you should carefully test and measure the performance impact.
```

Comment:
Batching is quite often the solution in the IEnumerable->IObservable bridge. Either async serial, or parallel.